### ISSUE TITLE
Terms and conditions - ask periodically to reaccept

### ISSUE THREAD

**Issue by cgeorgilakis** [keycloak/keycloak#19275]
### Description

We have a couple of cases (installations) in which we require users to periodically re-accept the Terms and Conditions since their last time they accepted them (i.e. should accept every 6 months since the last time). Currently, in keycloak there is no way to set such a policy. We have added this functionality to keycloak, so an admin can set such a periodic action. We have implemented it for the "terms and conditions" scenario and we would like to share this with the community. This can easily extended to include periodic reset of other user required actions, as for instance, the password.
We submit a pull request (which complies to all the rules), and we would be happy to make further changes to extend it for any other required actions you might want.

**Comment by jonkoops** [comment]
@cgeorgilakis is there a functionality currently that can trigger the user to re-accept the terms and conditions manually? Or would this have to be added to Keycloak?

**Comment by cgeorgilakis** [comment]
> @cgeorgilakis is there a functionality currently that can trigger the user to re-accept the terms and conditions manually? Or would this have to be added to Keycloak?

@jonkoops This has to be added to Keycloak (at least version 22). I do not know if any change related to T&C has been added after version 22. If not, could we open a github discussion for discussing the way we could do it?

**Comment by jonkoops** [comment]
I think we should think about how to implement this in a very minimal manner so it doesn't get stuck in review too much. 

Specifically I am thinking of adding a functionality that allows a manual re-trigger of the acceptance of the terms and conditions, so that doing so periodically can be done either through scripting, an extension to Keycloak, or Keycloak itself in the future.

WDYT?

**Comment by cgeorgilakis** [comment]
Manual re-trigger of the acceptance of the terms and conditions is [another issue related to T&C](https://issues.redhat.com/browse/KEYCLOAK-17379) that we have opened. It would be good to start implementing at least one of them.

**Comment by jonkoops** [comment]
I think that would be a good place to start working towards an implementation there. It can serve as a foundation for this feature. Could you log an issue for that on GitHub as well so we can track it?

**Comment by cgeorgilakis** [comment]
> I think that would be a good place to start working towards an implementation there. It can serve as a foundation for this feature. Could you log an issue for that on GitHub as well so we can track it?

https://github.com/keycloak/keycloak/issues/26124

**Comment by jonkoops** [comment]
Now that we know that there is a feature that forces users to re-accept the terms of service (see https://github.com/keycloak/keycloak/issues/26124#issuecomment-1888836345). I think we can focus on what this feature should look like based on @xianli123's design on the previous PR (https://github.com/keycloak/keycloak/pull/7895#issuecomment-1691137326):

![A screenshot of the new design](https://github.com/keycloak/keycloak/assets/695720/6b50492e-44c1-474f-8959-28026ab6bb49)

This adds a new feature, which is to allow uploading 'Terms and Conditions', whereas this is currently handled by adding [files to a theme](https://www.keycloak.org/docs/latest/server_admin/#proc-enabling-terms-conditions_server_administration_guide). I think this is the right way to go, as it is a lot more convenient than managing this inside of a theme. A couple of notes on this:

- I don't think this should be a file upload, but rather a code editor that allows you to add HTML.
- It is a common use-case that the 'Terms and Conditions' change based on locale, we should incorporate that here as well (perhaps a later enhancement).
- We need to decide what to do with the existing method of adding 'Terms and Conditions' through a template. Should we keep this? Deprecate the functionality?

As for the rest of the functionality, I find the form to become quite confusing. Why do we need to do this periodically? Should the user not be forced to accept the 'Terms and Conditions' every time they are changed?

I also believe we are not keeping track in the database of when the user last accepted the 'Terms and Conditions', so that would make this feature hard to implement without adding an extra field, or simply forcing all users to re-accept, even though they might already have done so recently.

@ssilvert @cgeorgilakis @xianli123 WDYT?

**Comment by xianli123** [comment]
Hi @jonkoops I am not sure what details should be covered by the `terms and conditions`. The original design is just based on the original issue. We need more elaborate details to do the design. Once we have the final requirements and concrete ideas, I can make a more reasonable design for this.

**Comment by ssilvert** [comment]
@jonkoops A required action is set until the user had done the action.  So we don't need an extra field to tell if the user has accepted T&C.  Once T&C is accepted, the required action is automatically unassigned from the user.

I assume there are legal reasons for forcing a user to re-accept T&C periodically.  Maybe @cgeorgilakis can confirm that this is the reason?

Adding the ability to upload a T&C document from the UI would be nice, but we just need to see if the core and store teams want to implement it.

The other related issue is #26124 where all users are forced to re-accept whenever the T&C changes.  This would just mean going through all the users and assigning the T&C required action.  It should probably be a manual process where the admin presses a button to update all users.

So what we really have is a more generic proposal for functionality that could apply to any required action:
1. Periodically set a required action.
2. Set a required action immediately for all users on demand via a button that the admin pushes.

We need the core and store teams to decide if we should go ahead with these before we start designing UI.

**Comment by jonkoops** [comment]
> A required action is set until the user had done the action. So we don't need an extra field to tell if the user has accepted T&C. Once T&C is accepted, the required action is automatically unassigned from the user.

I understand that, but it does not store **when** these terms were last accepted, only that they were accepted. So it would not be possible to say "All users that have accepted the terms before the last 6 months should be required to accept them again.". Perhaps this information is stored somewhere that I am not aware of though. 

I am a bit cautious of even introducing this feature to periodically re-accept the terms, as this sounds more like a band-aid to get around the problem that one does not know when the terms have changed, but @cgeorgilakis can attest to his use-case. So I am more inclined to go for what you mention as option 2.

Perhaps this could be a more generic feature to perform mass assignments of required actions to all users. This is also useful for other cases, such as forcing users to reset their password periodically, or when old credentials are somehow compromised.

**Comment by cgeorgilakis** [comment]
> I assume there are legal reasons for forcing a user to re-accept T&C periodically. Maybe @cgeorgilakis can confirm that this is the reason?

@jonkoops you know when user last time accept T&C (and last time every required action has been executed).
It is saved as user attribute in database (name= _terms_and_conditions_). See our declined [PR code](https://github.com/keycloak/keycloak/pull/7895/files#diff-7a78ab8c4b93bd2bcb9e7b2db7000aa21da3a2057a89d6070b778b38e5128235).
The strange thing is that we keep using  _terms_and_conditions_ even if now _TERMS_AND_CONDITIONS_ is new alias of required action.

**Comment by jonkoops** [comment]
So is there a use-case for doing this periodically, rather than as a mass action on all users whenever the terms are updated?

**Comment by cgeorgilakis** [comment]
@ssilvert @jonkoops we want to support both use cases (Periodically reaccept and mass action).
Periodically reaccept is needed for policy & legal reasons.

**Comment by jonkoops** [comment]
> Periodically reaccept is needed for policy & legal reasons.

That seems very vague, I can understand one needs to accept terms when they change, but what legal reason could you have to re-request this when it didn't change at all?

**Comment by cgeorgilakis** [comment]
> > Periodically reaccept is needed for policy & legal reasons.
> 
> That seems very vague, I can understand one needs to accept terms when they change, but what legal reason could you have to re-request this when it didn't change at all?

You could see our main client use case for periodically renewal in section 5.5 of [Community Membership Management Policy ](chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://documents.egi.eu/public/RetrieveFile?docid=3234&filename=EGI-SPG-Community-Member-Manage-V1.pdf&version=3).

I would like to mention that periodically reaccept is used in other well-known products.
For example you could see:
- section 13 of [Lab 23: Add terms of use and acceptance reporting of Microsoft Learn](https://microsoftlearning.github.io/SC-300-Identity-and-Access-Administrator/Instructions/Labs/Lab_23_AddTermsOfUseAcceptanceReporting.html)
- In [Configure Terms of Use in Azure Active Directory](http://www.thatlazyadmin.com/2022/07/20/configure-terms-of-use-in-azure-active-directory/) is mentioned that : 

> Require employees or guests to accept your terms of use policy on a recurring schedule.
